java-test-samples/apigw-lambda-list-s3-buckets/README.md

```diff
@@ -0,0 +1,295 @@
+```
+
+```bash
+# run a unit test in a separate terminal
+apigw-lambda-list-s3-buckets$ aws lambda invoke --function-name "ListBucketsFunction" --endpoint-url "http://127.0.0.1:3001" --no-verify-ssl out.txt
+```
+
+[[top]](#api-gateway-to-lambda-to-list-s3-buckets)
+
+## Use the SAM API Gateway emulator
+The SAM CLI can also emulate your application's API. Use the `sam local start-api` to run the API locally on port 3000.
+
+```bash
+# start a local emulator for an API Gateway endpoint
+apigw-lambda-list-s3-buckets$ sam local start-api
+```
+
+```bash
+# make a request to the endpoint in a separate terminal
+apigw-lambda-list-s3-buckets$ curl http://localhost:3000/listBuckets
+```
+
+The SAM CLI reads the application template to determine the API's routes and the functions that they invoke. The `Events` property on each function's definition includes the route and method for each path.
+
+```yaml
+Events:
+HelloWorld:
+  Type: Api
+  Properties:
+    Path: /listBuckets
+    Method: get
+```
+
+[[top]](#api-gateway-to-lambda-to-list-s3-buckets)
+
+## Run a unit tests
+
+Tests are defined in the `apigw-lambda-list-s3-buckets/src/test` folder in this project.
+
+```bash
+apigw-lambda-list-s3-buckets$ mvn test
+```
+
+[[top]](#api-gateway-to-lambda-to-list-s3-buckets)
+
+## Run a unit test using a mock framework
+You can use mocking frameworks like [Mockito](https://site.mockito.org/) to mock the service calls that are being done in the Lambda function.
+[`AppWithMockTest.java`](./src/test/java/com/example/AppWithMockTest.java) covers this example:
+
+```java
+@ExtendWith(MockitoExtension.class)
+public class AppWithMockTest { 
+  private static S3Client s3Client;
+
+  @BeforeEach
+  public void setup() {
+    s3Client = mock(S3Client.class);
+    Bucket bucket = mock(Bucket.class);
+    when(bucket.name()).thenReturn("foo");
+    lenient().when(s3Client.listBuckets())
+      .thenReturn(ListBucketsResponse.builder().buckets(bucket).build());
+  }
+
+  @Test
+  public void successfulResponse() {
+    App app = new App(s3Client);
+    APIGatewayProxyResponseEvent result = app.handleRequest(null, null);
+    assertEquals(200, result.getStatusCode().intValue());
+    assertEquals("text/plain", result.getHeaders().get("Content-Type"));
+    String content = result.getBody();
+    assertNotNull(content);
+    assertTrue(content.length() > 0);
+    assertEquals("foo", content);
+  }
+}
+```
+
+[[top]](#api-gateway-to-lambda-to-list-s3-buckets)
+
+## Run integration tests against cloud resources
+In order to run integration tests, you can use [Testcontainers and Localstack](https://www.testcontainers.org/modules/localstack/) which needs Docker engine running.
+
+[`AppIntegrationTest.java`](./src/test/java/com/example/AppIntegrationTest.java) shows the exact implementation. Important sections to look at are:
+
+
+```java
+private static final DockerImageName localStackImage =
+  DockerImageName.parse("localstack/localstack:0.14.3");
```

Comment:
more recent ?

Reply:
**localstack/localstack:1.2.0** is available https://hub.docker.com/r/localstack/localstack/tags.

Reply:
Upgraded to localstack:1.2.0